builder_test.go

```diff
@@ -282,163 +190,426 @@ func testBuilder(t *testing.T, when spec.G, it spec.S) {
+						`[[processes]]`+"\n"+
+							`type = "B-type"`+"\n"+
+							`command = "B-cmd"`+"\n"+
+							`[[processes]]`+"\n"+
+							`type = "override-type"`+"\n"+
+							`command = "B-cmd"`+"\n",
+						filepath.Join(appDir, "launch-B-v2.toml"),
+					)
+					metadata, err := builder.Build()
+					if err != nil {
+						t.Fatalf("Unexpected error:\n%s\n", err)
+					}
+					if s := cmp.Diff(metadata, &lifecycle.BuildMetadata{
+						Processes: []launch.Process{
+							{Type: "A-type", Command: "A-cmd", BuildpackID: "A"},
+							{Type: "B-type", Command: "B-cmd", BuildpackID: "B"},
+							{Type: "override-type", Command: "B-cmd", BuildpackID: "B"},
 						},
-						{
-							Require:   lifecycle.Require{Name: "dep2", Version: "v4"},
-							Buildpack: lifecycle.Buildpack{ID: "B", Version: "v2"},
+						Buildpacks: []lifecycle.Buildpack{
+							{ID: "A", Version: "v1", API: "0.3", Homepage: "Buildpack A Homepage"},
+							{ID: "B", Version: "v2", API: "0.2"},
 						},
-						{
-							Require:   lifecycle.Require{Name: "dep2-replace", Version: "v8"},
-							Buildpack: lifecycle.Buildpack{ID: "B", Version: "v2"},
+					}); s != "" {
+						t.Fatalf("Unexpected metadata:\n%s\n", s)
+					}
+				})
+
+				it("should return build metadata when processes are not present", func() {
+					metadata, err := builder.Build()
+					if err != nil {
+						t.Fatalf("Unexpected error:\n%s\n", err)
+					}
+					if s := cmp.Diff(metadata, &lifecycle.BuildMetadata{
+						Processes: []launch.Process{},
+						Buildpacks: []lifecycle.Buildpack{
+							{ID: "A", Version: "v1", API: "0.3", Homepage: "Buildpack A Homepage"},
+							{ID: "B", Version: "v2", API: "0.2"},
 						},
-					},
-				}); s != "" {
-					t.Fatalf("Unexpected:\n%s\n", s)
-				}
+					}); s != "" {
+						t.Fatalf("Unexpected:\n%s\n", s)
+					}
+				})
+			})
 
-				testPlan(t,
-					[]lifecycle.Require{
-						{Name: "dep1", Version: "v1"},
-						{Name: "dep1-next", Version: "v2"},
-						{Name: "dep1-replace", Version: "v3"},
-					},
-					filepath.Join(appDir, "build-plan-in-A-v1.toml"),
-				)
+			when("bom", func() {
+				when("platform api < 0.4", func() {
+					it.Before(func() {
+						builder.PlatformAPI = api.MustParse("0.3")
+					})
 
-				testPlan(t,
-					[]lifecycle.Require{
-						{Name: "dep1-next", Version: "v2"},
-						{Name: "dep2", Version: "v4"},
-						{Name: "dep2-next", Version: "v5"},
-						{Name: "dep2-replace", Version: "v6"},
-					},
-					filepath.Join(appDir, "build-plan-in-B-v2.toml"),
-				)
-			})
+					it("converts metadata.version to top level version", func() {
+						mkfile(t,
+							"[[entries]]\n"+
+								`name = "dep1"`+"\n"+
+								"[entries.metadata]\n"+
+								`version = "v1"`+"\n",
+							filepath.Join(appDir, "build-plan-out-A-v1.toml"),
+						)
+						metadata, err := builder.Build()
+						if err != nil {
+							t.Fatalf("Unexpected error:\n%s\n", err)
+						}
+						if s := cmp.Diff(metadata.BOM, []lifecycle.BOMEntry{
```

Comment:
Only asserting on the `BOM` here, whereas previously we also looked at `Processes` and `Buildpacks`. I think this makes it clearer but maybe the other way is safer.

Reply:
I support the decision to prioritize clarity.